parsley-debug/shared/src/main/scala/parsley/debugger/util/Collector.scala

```diff
@@ -0,0 +1,101 @@
+    def names(obj: Any): Unit = {
+        collectDefault() // Runs only once, ever, for a program execution.
+        Rename.addNames(XCollector.collectNames(obj))
+    }
+
+    /** Collect names of parsers from a [[parsley.token.Lexer]]. */
+    def lexer(lexer: Lexer): Unit = {
+        collectDefault()
+        Rename.addNames(XCollector.collectLexer(lexer))
+    }
+
+    // $COVERAGE-OFF$
+    /** Manually add a name for a parser by reference.
+      *
+      * Can also be used if a more informative name for a parser is wanted.
+      * In this case, use this method after using [[names]] or [[lexer]] to override the automatically
+      * collected / found name.
+      *
+      * @note Names assigned using this will take precedence over names assigned using [[parsley.debugger.combinator.named]].
+      */
+    def assignName(par: Parsley[_], name: String): Unit =
+        Rename.addName(par.internal, name)
+
+    /** Does the implementation of the collector for the current Scala platform actually work in
+      * automatically finding parsers in objects and getting their field names as written in your
+      * parser code?
+      *
+      * @note Manually named parsers using [[assignName]] or [[parsley.debugger.combinator.named]]
+      *       will still work regardless if the platform is supported or not.
+      */
+    @inline def isSupported: Boolean =
+        XCollector.supported
+
+    /** Collect the names of Parsley's various default singleton parsers. */
+    private var defaultCollected: Boolean = false
+    private def collectDefault(): Unit =
+        if (isSupported) {
+                this.synchronized {
+                if (!defaultCollected) {
+                    defaultCollected = true
+
+                    names(parsley.character)
+                    names(parsley.combinator)
+                    names(parsley.Parsley)
+                    names(parsley.position)
+                }
+            }
+        }
+    // $COVERAGE-ON$
+}
+
+/** A representation of the current implementation that [[Collector]] uses in order to
+  * actually collect the names of parsers. One of these will need to be implemented under the name
+  * `XCollector` under `parsley.debugger.util` for each different Scala runtime.
+  *
+  * @note This is an internal detail, so users do not have to interact with this if not necessary.
+  */
+// $COVERAGE-OFF$
+abstract class CollectorImpl private [parsley] () {
+    /** Collect names of parsers from an object. */
+    def collectNames(obj: Any): Map[LazyParsley[_], String]
+
+    /** Collect names of parsers from a [[parsley.token.Lexer]]. */
+    def collectLexer(lexer: Lexer): Map[LazyParsley[_], String]
+
+    /** Does the current platform's [[CollectorImpl]] actually get parsers from objects? */
+    val supported: Boolean
+
+    // Try grabbing a parser from a LazyParsley or Parsley instance.
+    protected def tryExtract(p: Any): LazyParsley[_] = {
```

Comment:
could this be something like:

```scala
p match {
  case p: LazyParsley[_] => p
  case p: Parsley[_] => p.internal
}
```

Reply:
(this does force a `Parsley[_]` object to be allocated though, I think)

Reply:
According to [the Scala 2 documentation](https://docs.scala-lang.org/overviews/core/value-classes.html#when-allocation-is-necessary) it would force a `Parsley[_]` object, I think.

Reply:
Although, I don't think the way I do it avoids that either?

Reply:
Swapped over to match as it avoids `try...catch` in [4842bec](https://github.com/j-mie6/parsley/pull/219/commits/4842bec37fb846b46458a17a4afd604abf03b22b)